router/protocol.go

```diff
@@ -2,58 +2,329 @@ package router
 
 import (
 	"bytes"
+	"encoding/gob"
+	"encoding/hex"
 	"fmt"
 	"io"
+	"time"
 )
 
 const (
-	Protocol                = "weave"
-	ProtocolVersion    byte = 1
-	ProtocolMinVersion byte = ProtocolVersion
+	Protocol           = "weave"
+	ProtocolMinVersion = 1
+	ProtocolMaxVersion = 2
 )
 
 var (
-	ProtocolBytes     = []byte(Protocol)
-	ProtocolLen       = len(ProtocolBytes)
-	ProtocolHeader    = append(ProtocolBytes, ProtocolMinVersion, ProtocolVersion)
-	ProtocolHeaderLen = len(ProtocolHeader)
+	ProtocolBytes = []byte(Protocol)
+
+	HeaderTimeout = 10 * time.Second
+
+	ProtocolV1Features = []string{
+		"ConnID",
+		"Name",
+		"NickName",
+		"PeerNameFlavour",
+		"UID",
+	}
+
+	ErrExpectedCrypto   = fmt.Errorf("Password specified, but peer requested an unencrypted connection")
+	ErrExpectedNoCrypto = fmt.Errorf("No password specificed, but peer requested an encrypted connection")
 )
 
-func exchangeProtocolHeader(w io.Writer, r io.Reader) (byte, error) {
-	if n, err := w.Write(ProtocolHeader); err != nil {
-		return 0, err
-	} else if n != ProtocolHeaderLen {
-		return 0, fmt.Errorf("failed to send complete protocol header")
+// We don't need the full net.TCPConn to do the protocol intro.  This
+// interface contains just the parts we do need, to support testing
+type ProtocolIntroConn interface {
+	// io.Reader
+	Read(b []byte) (n int, err error)
+
+	// io.Writer
+	Write(b []byte) (n int, err error)
+
+	// net.Conn's deadline methods
+	SetDeadline(t time.Time) error
+	SetReadDeadline(t time.Time) error
+	SetWriteDeadline(t time.Time) error
+}
+
+type ProtocolIntroParams struct {
+	MinVersion byte
+	MaxVersion byte
+	Features   map[string]string
+	Conn       ProtocolIntroConn
+	Password   []byte
+	Outbound   bool
+}
+
+type ProtocolIntroResults struct {
+	Features   map[string]string
+	Receiver   TCPReceiver
+	Sender     TCPSender
+	SessionKey *[32]byte
+	Version    byte
+}
```

Comment:
This was deliberate.  I wanted to restrict the code in `protocol.go` to dealing with the protocol only, and exclude the interpretation of the data passed over the protocol (this line is subjective, but the result seems quite clean to me).  And I regard the `features` type as just a utility for the sake of `parseFeatures` (and possibly for other future uses of the features in connection-related code).  So it would seem out of place if hoisted into `protocol.go`.